File: camerax_android/android/src/main/kotlin/dev/hebei/camerax_android/legacy/common/PermissionManager.kt
package dev.hebei.camerax_android.legacy.common

import dev.hebei.camerax_android.common.PermissionManager
import dev.hebei.camerax_android.legacy.CameraXRegistrar
import dev.hebei.camerax_android.legacy.Permission
import dev.hebei.camerax_android.legacy.PigeonApiPermissionManager
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class PermissionManager(registrar: CameraXRegistrar) : PigeonApiPermissionManager(registrar) {
    override fun instance(): PermissionManager {
        return PermissionManager
    }

    override fun checkPermissioin(
        pigeon_instance: PermissionManager, permission: Permission
    ): Boolean {
        return pigeon_instance.checkPermission(permission.obj)
    }

    override fun requestPermissions(
        pigeon_instance: PermissionManager,
        permissions: List<Permission>,
        callback: (Result<Boolean>) -> Unit
    ) {
        CoroutineScope(Dispatchers.Main).launch {
            try {
                val permissionObjs = permissions.map { permission -> permission.obj }
                val isGranted = pigeon_instance.requestPermissions(permissionObjs)
                callback(Result.success(isGranted))
            } catch (e: Exception) {
                callback(Result.failure(e))
            }
        }
    }
}

val Permission.obj
    get() = when (this) {
        Permission.ALBUM -> PermissionManager.Permission.ALBUM
        Permission.AUDIO -> PermissionManager.Permission.AUDIO
        Permission.VIDEO -> PermissionManager.Permission.VIDEO
    }

File: camerax/lib/src/common/permission_manager.dart
import 'package:camerax_platform_interface/camerax_platform_interface.dart'
    as $interface;

class PermissionManager {
  final $interface.PermissionManager obj;

  PermissionManager() : obj = $interface.PermissionManager();

  Future<bool> checkPermission($interface.Permission permission) async {
    final isGranted = await obj.checkPermission(permission);
    return isGranted;
  }

  Future<bool> requestPermissions(
      List<$interface.Permission> permissions) async {
    final isGranted = await obj.requestPermissions(permissions);
    return isGranted;
  }
}
